src/rai/rai/communication/ros2/api.py

```diff
@@ -0,0 +1,398 @@
+            reliability_reliable_count += 1
+        if profile.durability == DurabilityPolicy.TRANSIENT_LOCAL:
+            durability_transient_local_count += 1
+
+    request_qos = QoSProfile(
+        history=HistoryPolicy.KEEP_LAST,
+        depth=1,
+        liveliness=LivelinessPolicy.AUTOMATIC,
+    )
+
+    # Set reliability based on publisher offers
+    if reliability_reliable_count == num_endpoints:
+        request_qos.reliability = ReliabilityPolicy.RELIABLE
+    else:
+        if reliability_reliable_count > 0:
+            logging.warning(
+                "Some, but not all, publishers are offering RELIABLE reliability. "
+                "Falling back to BEST_EFFORT as it will connect to all publishers. "
+                "Some messages from Reliable publishers could be dropped."
+            )
+        request_qos.reliability = ReliabilityPolicy.BEST_EFFORT
+
+    # Set durability based on publisher offers
+    if durability_transient_local_count == num_endpoints:
+        request_qos.durability = DurabilityPolicy.TRANSIENT_LOCAL
+    else:
+        if durability_transient_local_count > 0:
+            logging.warning(
+                "Some, but not all, publishers are offering TRANSIENT_LOCAL durability. "
+                "Falling back to VOLATILE as it will connect to all publishers. "
+                "Previously-published latched messages will not be retrieved."
+            )
+        request_qos.durability = DurabilityPolicy.VOLATILE
+
+    return request_qos
+
+
+def build_ros2_msg(msg_type: str, msg_args: Dict[str, Any]) -> object:
+    """Build a ROS2 message instance from type string and content dictionary."""
+    msg_cls = import_message_from_str(msg_type)
+    msg = msg_cls()
+    rosidl_runtime_py.set_message.set_message_fields(msg, msg_args)
+    return msg
+
+
+def build_ros2_service_request(
+    service_type: str, service_request_args: Dict[str, Any]
+) -> Tuple[object, Type[Any]]:
+    msg_cls = import_message_from_str(service_type)
+    msg = msg_cls.Request()
+    rosidl_runtime_py.set_message.set_message_fields(msg, service_request_args)
+    return msg, msg_cls  # type: ignore
+
+
+class ROS2TopicAPI:
+    """Handles ROS2 topic operations including publishing and subscribing to messages.
+
+    This class provides a high-level interface for ROS2 topic operations with automatic
+    QoS profile matching and proper resource management.
+
+    Attributes:
+        node: The ROS2 node instance
+        logger: Logger instance for this class
+        _publishers: Dictionary mapping topic names to their publisher instances
+    """
+
+    def __init__(self, node: rclpy.node.Node) -> None:
+        """Initialize the ROS2 topic API.
+
+        Args:
+            node: ROS2 node instance to use for communication
+        """
+        self._node = node
+        self._logger = node.get_logger()
+        self._publishers: Dict[str, Publisher] = {}
+
+    def list_topics(self) -> List[Tuple[str, List[str]]]:
```

Comment:
If it's a shortcut, maybe better to keep the name of the method consistent with the original?
```suggestion
    def get_topic_names_and_types(self) -> List[Tuple[str, List[str]]]:
```

Reply:
Done, chaned it a bit